lib/charms/alertmanager_k8s/v0/alertmanager_dispatch.py

```diff
@@ -0,0 +1,257 @@
+    then a :class:`ClusterChanged` event should be emitted.
+    """
+
+
+class AlertmanagerConsumerEvents(ObjectEvents):
+    """Event descriptor for events raised by `AlertmanagerConsumer`."""
+
+    cluster_changed = EventSource(ClusterChanged)
+
+
+class RelationManagerBase(Object):
+    """Base class that represents relation ends ("provides" and "requires").
+
+    :class:`RelationManagerBase` is used to create a relation manager. This is done by inheriting
+    from :class:`RelationManagerBase` and customising the sub class as required.
+
+    Attributes:
+        name (str): consumer's relation name
+    """
+
+    def __init__(self, charm: CharmBase, relation_name):
+        super().__init__(charm, relation_name)
+        self.name = relation_name
+
+
+class AlertmanagerConsumer(RelationManagerBase):
+    """A "consumer" handler to be used by charms that relate to Alertmanager (the 'requires' side).
+
+    To have your charm consume alertmanager cluster data, declare the interface's use in your
+    charm's metadata.yaml file:
+
+    ```yaml
+    requires:
+      alertmanager:
+        interface: alertmanager_dispatch
+    ```
+
+    A typical example of importing this library might be
+
+    ```python
+    from charms.alertmanager_k8s.v0.alertmanager_dispatch import AlertmanagerConsumer
+    ```
+
+    In your charm's `__init__` method:
+
+    ```python
+    self.alertmanager_consumer = AlertmanagerConsumer(self, relation_name="alertmanager")
+    ```
+
+    Every change in the alertmanager cluster emits a :class:`ClusterChanged` event that the
+    consumer charm can register and handle, for example:
+
+    ```
+    self.framework.observe(self.alertmanager_consumer.on.cluster_changed,
+                           self._on_alertmanager_cluster_changed)
+    ```
+
+    The updated alertmanager cluster can then be obtained via the `get_cluster_info` method
+
+    This consumer library expect the consumer charm to observe the `cluster_changed` event.
+
+    Arguments:
+            charm (CharmBase): consumer charm
+            relation_name (str): from consumer's metadata.yaml
+
+    Attributes:
+            charm (CharmBase): consumer charm
+    """
+
+    on = AlertmanagerConsumerEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+
+        self.framework.observe(
+            self.charm.on[self.name].relation_changed, self._on_relation_changed
+        )
+        self.framework.observe(
+            self.charm.on[self.name].relation_departed,
+            self._on_relation_departed,
+        )
+        self.framework.observe(self.charm.on[self.name].relation_broken, self._on_relation_broken)
+
+    def _on_relation_changed(self, event: ops.charm.RelationChangedEvent):
+        """This hook notifies the charm that there may have been changes to the cluster."""
+        if event.unit:  # event.unit may be `None` in the case of app data change
```

Comment:
Is there any reason not to emit a cluster changed event when application data for the cluster changes ?

Reply:
@sed-i 👆

Reply:
In the general case we could. In the context of alertmanager, "cluster changed" is only sensible for changes in unit data. At least for now. The `if event.unit` reduces false positives / event storm. The `if event.unit` is not intended for the entire `_on_relation_changed` hook, but rather for `on.cluster_changed.emit()` only.